### PR TITLE
Week4

### PR DESCRIPTION
**Pull request by shiv123-max**
The questions are there in week4 folder in PDF format.

### CODE REVIEW

File: Week-4/4.1- Promises/index.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+const getNumber = () => {
+  const randomNumber = Math.floor(Math.random() * 1000);
+  console.log("Random number is:", randomNumber);
+  if (randomNumber % 5 == 0) {
```
Comment (kapilggg10) [inline Week-4/4.1- Promises/index.js:4]: `return randomNumber % 5 === 0`

can use this to avoid if-else

File: Week-4/4.1- Promises/index.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+const REJECTED = 2;
+
+function MyPromise(executor) {
+  let state = PENDING;
+  let value = null;
+  let handlers = [];
+  let catchers = [];
+
+  function resolve(val) {
+    if (state !== PENDING) return;
```
Comment (kapilggg10) [inline Week-4/4.1- Promises/index.js:22]: good 👍

File: Week-4/4.1- Promises/index.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        reject("Divisible by 5");
+      }, 2000)
+    : setTimeout(() => {
+        resolve("Not Divisible by 5");
+      }, 3000);
+});
+
+p.then((val) => console.log("Fulfilled ->", val)).catch((err) =>
+  console.log("Rejected ->", err)
+);
```
Comment (kapilggg10) [inline Week-4/4.1- Promises/index.js:70]: when we do chaining of promises, order is not maintained of execution.
`p
    .then((val) => 6)
    .catch((err) => console.log("Rejected ->", err)).
    then((val) => console.log('hello', val))
    .then(() => console.log('hell'));`

File: Week-4/4.1- Promises/index.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  function resolve(val) {
+    if (state !== PENDING) return;
+    value = val;
+    state = FULFILLED;
+
+    handlers.forEach((callback) => callback(value));
+  }
+
+  function reject(val) {
```
Comment (kapilggg10) [inline Week-4/4.1- Promises/index.js:29]: should have this.bind method as mentioned in assignment.

File: Week-4/4.1- Promises/index.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      catchers.push(failureCallback);
+    }
+    return this;
+  };
+
+  executor(resolve, reject);
+}
+
+const p = new MyPromise((resolve, reject) => {
+  getNumber()
```
Comment (kapilggg10) [inline Week-4/4.1- Promises/index.js:59]: should create getNumber using your new MyPromise

Comment (shiv123-max) [reply Week-4/4.1- Promises/index.js:59]: Sir, in the question it is written that implement getNumber function which generates a random number. It is not written to create getNumber using promises. It says reject promise if divisible by 5. What should I do?

Comment (kapilggg10) [reply Week-4/4.1- Promises/index.js:59]: It's ok, I meant you could implement that as a promise that could have been a better approach to implement this, but not necessary.

**Pull request merged**